File: problems/038-conformance-feature-coverage/solutions/python/main.py
import sys

t = list(map(int, sys.stdin.buffer.read().split()))
it = iter(t)
F, N, B = next(it), next(it), next(it)
S = 1 << F
inf = 10**30
dp = [inf] * S
dp[0] = 0
for _ in range(N):
    cost, k = next(it), next(it)
    m = 0
    for _ in range(k):
        m |= 1 << (next(it) - 1)
    ndp = dp[:]
    for s, v in enumerate(dp):
        if v != inf:
            q = s | m
            ndp[q] = min(ndp[q], v + cost)
    dp = ndp
print(max(s.bit_count() for s, v in enumerate(dp) if v <= B))

File: problems/038-conformance-feature-coverage/solutions/rust/main.rs
use std::io::{self, Read};
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut t = s.split_whitespace();
    let f: usize = t.next().unwrap().parse().unwrap();
    let n: usize = t.next().unwrap().parse().unwrap();
    let b: u64 = t.next().unwrap().parse().unwrap();
    let states = 1usize << f;
    let inf = u64::MAX / 4;
    let mut dp = vec![inf; states];
    dp[0] = 0;
    for _ in 0..n {
        let cost: u64 = t.next().unwrap().parse().unwrap();
        let k: usize = t.next().unwrap().parse().unwrap();
        let mut m = 0usize;
        for _ in 0..k {
            let x: usize = t.next().unwrap().parse().unwrap();
            m |= 1 << (x - 1);
        }
        let mut next = dp.clone();
        for mask in 0..states {
            if dp[mask] != inf {
                let q = mask | m;
                next[q] = next[q].min(dp[mask] + cost);
            }
        }
        dp = next;
    }
    let mut ans = 0;
    for mask in 0..states {
        if dp[mask] <= b {
            ans = ans.max(mask.count_ones());
        }
    }
    println!("{}", ans);
}
